File: frontend/src/types/product.ts
export interface ProductItem {
  service_product_code: string
  service_product_name: string
  product_type: string
  category: string
  service_order_type: string
  unit: string
  price: string
  price_status: string
  related_category: string
  related_area: string
  fault_phenomenon: string
  remark: string
}

export interface ProductSearchResult {
  score: number
  service_product_code: string
  service_product_name: string
  service_order_type: string
  product_type: string
  related_area: string
  fault_phenomenon: string
  price: string
  unit: string
}

export interface ProductSearchDiagnosticCandidate {
  service_product_code?: string | null
  service_product_name?: string | null
  service_order_type?: string | null
  fault_phenomenon?: string | null
  vector_score?: number | null
  keyword_overlap?: boolean | null
  fault_keyword_overlap?: boolean | null
  penalty?: number | null
  bonus?: number | null
  adjusted_score?: number | null
  included?: boolean
  filtered_reason?: string | null
}

export interface ProductSearchDiagnostics {
  query: string
  top_k?: number | null
  threshold?: number | null
  has_fault?: boolean
  fetch_k?: number | null
  fallback_to_vector_results?: boolean
  returned_count?: number | null
  returned_codes?: string[]
  reason?: string | null
  candidates?: ProductSearchDiagnosticCandidate[]
}

export interface ProductSearchResponse {
  query: string
  count: number
  products: ProductSearchResult[]
  diagnostics?: ProductSearchDiagnostics | null
}

File: schemas/product.py
from pydantic import BaseModel, Field


class ProductItem(BaseModel):
    service_product_code: str
    service_product_name: str
    product_type: str
    category: str
    service_order_type: str
    unit: str
    price: str
    price_status: str
    related_category: str
    related_area: str
    fault_phenomenon: str
    remark: str


class ProductListResponse(BaseModel):
    total: int
    items: list[ProductItem]


class ProductSearchRequest(BaseModel):
    query: str
    top_k: int = Field(default=10, ge=1, le=50)
    threshold: float | None = Field(default=None, ge=0.0, le=1.0)


class ProductSearchResult(BaseModel):
    score: float
    service_product_code: str
    service_product_name: str
    service_order_type: str
    product_type: str
    related_area: str
    fault_phenomenon: str
    price: str
    unit: str


class ProductSearchResponse(BaseModel):
    query: str
    count: int
    products: list[ProductSearchResult]
